aws-fargate/miniwdl_aws_fargate.py

```diff
@@ -0,0 +1,208 @@
+
+    @classmethod
+    def global_init(cls, cfg: config.Loader, logger: logging.Logger) -> None:
+        try:
+            cls.efs_security_group = cfg["aws_fargate"]["efs_security_group"]
+        except config.ConfigMissing:
+            cls.efs_security_group = cls.default_efs_security_group
+        try:
+            cls.efs_id = cfg["aws_fargate"]["efs_id"]
+        except config.ConfigMissing:
+            cls.efs_id = None
+        try:
+            cls.efs_mountpoint = cfg["aws_fargate"]["efs_mountpoint"]
+        except config.ConfigMissing:
+            cls.efs_mountpoint = None
+
+        for partition in psutil.disk_partitions(all=True):
+            if partition.fstype == "nfs4":
+                if f".efs.{ecs.clients.ecs.meta.region_name}.amazonaws.com:/" in partition.device:
+                    # if os.stat(host_dir).st_dev == os.stat(partition.mountpoint).st_dev:
+                    if cls.efs_id is None:
+                        cls.efs_id = partition.device.split(":")[0].split(".")[-5]
+                    elif cls.efs_id not in partition.device:
+                        msg = "Expected filesystem mount {} to match configured EFS ID {}"
+                        raise RuntimeError(msg.format(partition, cls.efs_id))
+                    if cls.efs_mountpoint is None:
+                        cls.efs_mountpoint = partition.mountpoint
+                    elif cls.efs_mountpoint != partition.mountpoint:
+                        msg = "Expected filesystem mount {} to match configured EFS mountpoint {}"
+                        raise RuntimeError(msg.format(partition, cls.efs_mountpoint))
+                    return
+        else:
+            if cls.efs_id is None:
+                raise RuntimeError(
+                    'EFS filesystem config is missing. Mount the filesystem and run miniwdl in a subdirectory of the '
+                    'mountpoint or set the filesystem ID with "export MINIWDL__AWS_FARGATE__EFS_ID=fs-12345678 '
+                    'MINIWDL__AWS_FARGATE__EFS_MOUNTPOINT=/mnt/efs" or any other miniwdl config facility. '
+                    'Use "aws efs describe-file-systems" to list filesystems.'
+                )
+            logger.info("Mounting %s at %s", cls.efs_id, cls.efs_mountpoint)
+            fs_url = f"{cls.efs_id}.efs.{ecs.clients.ecs.meta.region_name}.amazonaws.com:/"
+            subprocess.run(["sudo", "mount", "-t", "nfs4", "-o",
+                            "nfsvers=4.1,rsize=1048576,wsize=1048576,hard,timeo=600,retrans=2",
+                            fs_url, cls.efs_mountpoint])
+
+    @classmethod
+    def detect_resource_limits(cls, cfg: config.Loader, logger: logging.Logger) -> Dict[str, int]:
+        return cls._limits
+
+    def poll_task(
+        self, logger: logging.Logger, task_desc, verbose: bool = False
+    ) -> Optional[int]:
+        res = ecs.clients.ecs.describe_tasks(cluster=task_desc["clusterArn"], tasks=[task_desc["taskArn"]])
```

Comment:
Polling the swarm daemon is cheap because it's typically a unix socket request to the local daemon. I wonder if we might eventually need something better for ECS if we're operating dozens or hundreds of jobs, like maybe some other background thread to do a batch DescribeTasks periodically.

Reply:
I agree, the client will get rate limited quickly when we scale this up. We need a batching/queueing abstraction here, but I'll hold off on that and add it separately.

Reply:
Filed #5 to track this.